handler/src/main/java/io/netty/handler/ssl/JettyAlpnSslEngine.java

```diff
@@ -28,45 +29,59 @@
-        if (available || PlatformDependent.javaVersion() > 8) {
-            return;
+        if (PlatformDependent.javaVersion() <= 8) {
+            try {
+                // Always use bootstrap class loader.
+                Class.forName("sun.security.ssl.ALPNExtension", true, null);
+                return true;
+            } catch (Throwable ignore) {
+                // alpn-boot was not loaded.
+            }
         }
+        return false;
+    }
 
-        try {
-            // Always use bootstrap class loader.
-            Class.forName("sun.security.ssl.ALPNExtension", true, null);
-            available = true;
-        } catch (Exception ignore) {
-            // alpn-boot was not loaded.
-        }
+    static JettyAlpnSslEngine newClientEngine(SSLEngine engine,
+            JdkApplicationProtocolNegotiator applicationNegotiator) {
+        return new ClientEngine(engine, applicationNegotiator);
     }
 
-    JdkAlpnSslEngine(SSLEngine engine, final JdkApplicationProtocolNegotiator applicationNegotiator, boolean server) {
+    static JettyAlpnSslEngine newServerEngine(SSLEngine engine,
+            JdkApplicationProtocolNegotiator applicationNegotiator) {
+        return new ServerEngine(engine, applicationNegotiator);
+    }
+
+    private JettyAlpnSslEngine(SSLEngine engine) {
         super(engine);
-        checkNotNull(applicationNegotiator, "applicationNegotiator");
+    }
 
-        if (server) {
-            final ProtocolSelector protocolSelector = checkNotNull(applicationNegotiator.protocolSelectorFactory()
-                    .newSelector(this, new LinkedHashSet<String>(applicationNegotiator.protocols())),
-                    "protocolSelector");
-            ALPN.put(engine, new ServerProvider() {
+    private static final class ClientEngine extends JettyAlpnSslEngine {
+        ClientEngine(SSLEngine engine, final JdkApplicationProtocolNegotiator applicationNegotiator) {
+            super(engine);
+            checkNotNull(applicationNegotiator, "applicationNegotiator");
+            final ProtocolSelectionListener protocolListener = checkNotNull(applicationNegotiator
+                            .protocolListenerFactory().newListener(this, applicationNegotiator.protocols()),
+                    "protocolListener");
+            ALPN.put(engine, new ALPN.ClientProvider() {
                 @Override
-                public String select(List<String> protocols) throws SSLException {
+                public List<String> protocols() {
+                    return applicationNegotiator.protocols();
+                }
+
+                @Override
+                public void selected(String protocol) throws SSLException {
                     try {
-                        return protocolSelector.select(protocols);
+                        protocolListener.selected(protocol);
                     } catch (SSLHandshakeException e) {
```

Comment:
can we share the `toSSLHandshakeException` from `ConscryptAlpnSslEngine` and use it in this class? Maybe move it into a package private utility class somewhere (to avoid class loading attempting to load the native library).